Linking failure on ARM (Raspberry Pi 4, Raspbian 32-bit)

**libtorrent version (or branch)**

`1.2.10`

**platform/architecture**

ARM (Raspberry Pi 4b)
Rasbian Linux 32-bit

**compiler and compiler version**

- GCC 8.3.0

Libraries/tools:

- OpenSSL 1.1.1d
- Boost 1.74.0
- CMake 3.18.2

**please describe what symptom you see, what you would expect to see instead and
how to reproduce it.**

libtorrent builds successfully, but linking always fails with errors relating to `libatomic`. The simplest way to reproduce is to attempt to build `examples` and or `tools` (`-Dbuild_examples=ON`, `-Dbuild_tools=ON`).

This seems to be a widespread problem with Raspbian in general (just search for `raspberrypi libatomic` on the web for other examples).

Relevant snippet of the error:

```
[160/179] Linking CXX executable tools/session_log_alerts                                                                                                                                                                                                                
FAILED: tools/session_log_alerts                                                                                                                                                                                                                                 
: && /usr/bin/c++ -O3 -DNDEBUG  tools/CMakeFiles/session_log_alerts.dir/session_log_alerts.cpp.o -o tools/session_log_alerts  -Wl,-rpath,/home/pi/libtorrent/cmake-build-dir:/opt/boost-1.74.0/lib  libtorrent-rasterbar.so.1.2.10  -lpthread  
/usr/lib/arm-linux-gnueabihf/libssl.so  /usr/lib/arm-linux-gnueabihf/libcrypto.so  /opt/boost-1.74.0/lib/libboost_system.so.1.74.0 && :                                                                                                                                  
/usr/bin/ld: libtorrent-rasterbar.so.1.2.10: undefined reference to `__atomic_compare_exchange_8'                                                                                                                                                                        
/usr/bin/ld: libtorrent-rasterbar.so.1.2.10: undefined reference to `__atomic_load_8'                                                                                                                                                                                    
/usr/bin/ld: libtorrent-rasterbar.so.1.2.10: undefined reference to `__atomic_store_8'                                                                                                                                 
/usr/bin/ld: libtorrent-rasterbar.so.1.2.10: undefined reference to `__atomic_fetch_add_8'                                                                                                                            
collect2: error: ld returned 1 exit status
```

What would be the best way for libtorrent to deal with this? So far I have found the following workarounds, to at least build the examples and tools:

- Passing `-DCMAKE_EXE_LINKER_FLAGS="-latomic"` to the configure command line
- This patch:

    ```diff
    diff --git a/examples/CMakeLists.txt b/examples/CMakeLists.txt
    index d35d54152..e49b048c2 100644
    --- a/examples/CMakeLists.txt
    +++ b/examples/CMakeLists.txt
    @@ -12,6 +12,8 @@ set(single_file_examples
     foreach(example ${single_file_examples})
         add_executable(${example} "${example}.cpp")
         target_link_libraries(${example} PRIVATE torrent-rasterbar)
    +    # XXX: raspberry pi workaround
    +    target_link_libraries(${example} PRIVATE atomic)
     endforeach(example)
     
     add_executable(client_test
    @@ -20,3 +22,6 @@ add_executable(client_test
         torrent_view.cpp
         session_view.cpp)
     target_link_libraries(client_test PRIVATE torrent-rasterbar)
    +# XXX: raspberrry pi workaround
    +target_link_libraries(client_test PRIVATE atomic)
    +
    diff --git a/tools/CMakeLists.txt b/tools/CMakeLists.txt
    index 1bb1b5ea7..21344c799 100644
    --- a/tools/CMakeLists.txt
    +++ b/tools/CMakeLists.txt
    @@ -3,3 +3,7 @@ target_link_libraries(dht PRIVATE torrent-rasterbar)
     
     add_executable(session_log_alerts session_log_alerts.cpp)
     target_link_libraries(session_log_alerts PRIVATE torrent-rasterbar)
    +
    +# XXX: raspberrry pi workaround
    +target_link_libraries(dht PRIVATE atomic)
    +target_link_libraries(session_log_alerts PRIVATE atomic)
    ```

Example of another solution: https://github.com/pothosware/SoapyRTLSDR/pull/42

> But, as far as I can tell, this is an issue in GCC. There should be a work-around in cmake, and the fact that there isn't one, makes it an issue in cmake too (in my opinion).

True. Just tested with clang-9 (9.0.1), it does not suffer from this problem. Just a few seemingly harmless warnings:

```
[11/190] Building CXX object CMakeFiles/torrent-rasterbar.dir/src/cpuid.cpp.o                                                                                                                                        
../src/cpuid.cpp:75:9: warning: macro is not used [-Wunused-macros]
#define helper_getauxval getauxval               
        ^                       
1 warning generated.                                 
[124/190] Building CXX object CMakeFiles/torrent-rasterbar.dir/deps/try_signal/try_signal.cpp.o               
../deps/try_signal/try_signal.cpp:79:5: warning: disabled expansion of recursive macro [-Wdisabled-macro-expansion]
        sa.sa_sigaction = &sig::detail::handler;
           ^                                                       
/usr/include/arm-linux-gnueabihf/bits/sigaction.h:40:43: note: expanded from macro 'sa_sigaction'
# define sa_sigaction   __sigaction_handler.sa_sigaction       
                                            ^
1 warning generated.                    
[178/190] Building CXX object examples/CMakeFiles/connection_tester.dir/connection_tester.cpp.o
../examples/connection_tester.cpp:561:59: warning: implicit conversion changes signedness: 'unsigned int' to '__gnu_cxx::__normal_iterator<libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void> *, std::vector<libtorrent::aux::strong_typedef<in
t, libtorrent::aux::piece_index_tag, void>, std::allocator<libtorrent::aux::strong_typedef<int, libtorrent::aux::piece_index_tag, void> > > >::difference_type' (aka 'int') [-Wsign-conversion]
                                else pieces.insert(pieces.begin() + (unsigned(rand()) % pieces.size()), piece);
                                                                  ~  ~~~~~~~~~~~~~~~~~^~~~~~~~~~~~~~~
1 warning generated.                                 
[190/190] Linking CXX executable examples/client_test

```

--- a/CMakeLists.txt
+++ b/CMakeLists.txt
@@ -629,6 +629,38 @@ if (APPLE)
 	target_link_libraries(torrent-rasterbar PRIVATE "-framework CoreFoundation" "-framework SystemConfiguration")
 endif()
 
+# check if we need to link with libatomic (not needed on MSVC)
+if (NOT Windows)
+	# TODO: migrate to CheckSourceCompiles in CMake >= 3.19
+	include(CheckCXXSourceCompiles)
+
+	set(ATOMICS_TEST_SOURCE [=[
+		#include <atomic>
+		#include <cstdint>
+		std::atomic<int> x{0};
+		int main() {
+			x.fetch_add(1, std::memory_order_relaxed);
+			return 0;
+		}
+	]=])
+	string(REPLACE "std::atomic<int>" "std::atomic<std::int64_t>" ATOMICS64_TEST_SOURCE "${ATOMICS_TEST_SOURCE}")
+
+	check_cxx_source_compiles("${ATOMICS_TEST_SOURCE}" HAVE_CXX_ATOMICS_WITHOUT_LIB)
+	check_cxx_source_compiles("${ATOMICS64_TEST_SOURCE}" HAVE_CXX_ATOMICS64_WITHOUT_LIB)
+	if((NOT HAVE_CXX_ATOMICS_WITHOUT_LIB) OR (NOT HAVE_CXX_ATOMICS64_WITHOUT_LIB))
+		set(CMAKE_REQUIRED_LIBRARIES "atomic")
+		check_cxx_source_compiles("${ATOMICS_TEST_SOURCE}" HAVE_CXX_ATOMICS_WITH_LIB)
+		check_cxx_source_compiles("${ATOMICS64_TEST_SOURCE}" HAVE_CXX_ATOMICS64_WITH_LIB)
+		if ((NOT HAVE_CXX_ATOMICS_WITH_LIB) OR (NOT HAVE_CXX_ATOMICS64_WITH_LIB))
+			message(FATAL_ERROR "No native support for std::atomic, or libatomic not found!")
+		else()
+			message(STATUS "Linking with libatomic for atomics support")
+			unset(CMAKE_REQUIRED_LIBRARIES)
+			target_link_libraries(torrent-rasterbar PUBLIC atomic)
+		endif()
+	endif()
+endif()
+
 feature_option(build_tests "build tests" OFF)
 feature_option(build_examples "build examples" OFF)
 feature_option(build_tools "build tools" OFF)

